File: src/Algorithms/Kadanes_Algorithm/28-Find-the-Index-of-the-First-Occurrence-in-a-String.js
var strStr = function(haystack, needle) {
    const freqMap = Array.from({ length: 26}, () => 0);

    [...needle].forEach(ch => {
        freqMap[ch.charCodeAt(0) - 'a'.charCodeAt(0)] += 1;
    })

    let left = 0;

    for (let right = 0; right < haystack.length; right += 1) {
        
        freqMap[haystack.charCodeAt(right) - 'a'.charCodeAt(0)] -= 1;

        while (freqMap[haystack.charCodeAt(right) - 'a'.charCodeAt(0)] < 0) {
            freqMap[haystack.charCodeAt(left) - 'a'.charCodeAt(0)] += 1;
            left += 1;
        }

        if (haystack[left] === needle[0] && right - left + 1 === needle.length) {
            if (haystack.substring(left, right + 1) === needle) return left;
        } 
    }

    return -1;
}


// Kadane's algrithm (Template 2)
var strStr = function(haystack, needle) {
    if (needle === "") return 0;  // edge case for empty needle

    for (let left = 0; left <= haystack.length - needle.length; left++) {
        let right = 0;
        while (right < needle.length && haystack[left + right] === needle[right]) {
            right++;
        }
        if (right === needle.length) {
            return left;  // found the needle
        }
    }
    return -1;  // needle not found
};

File: src/Algorithms/Kadanes_Algorithm/28-Find-the-Index-of-the-First-Occurrence-in-a-String.java
class Solution {

    public int strStr(String haystack, String needle) {
        int[] freqCounter = new int[26];

        for (int i = 0; i < needle.length(); i++) {
            freqCounter[needle.charAt(i) - "a".charAt(0)]++;
        }

        int left = 0;

        for (int right = 0; right < haystack.length(); right++) {

            freqCounter[haystack.charAt(right) - "a".charAt(0)]--;

            while (freqCounter[haystack.charAt(right) - "a".charAt(0)] < 0) {
                freqCounter[haystack.charAt(left) - "a".charAt(0)]++;
                left++;
            }

            if (right - left + 1 == needle.length())
                return left;
        }

        return -1;
    }

    public int strStrV2(String haystack, String needle) {

        if (needle == "")
            return 0;

        char[] needleArr = needle.toCharArray();
        char[] haystackArr = haystack.toCharArray();

        for (int left = 0; left <= haystackArr.length - needleArr.length; left++) {
            int right = 0;
            while (right < needleArr.length && haystackArr[left + right] == needleArr[right]) {
                right++;
            }
            System.out.println(right);
            if (right == needleArr.length)
                return left;
        }

        return -1;
    }

}
